### PR TITLE
Resolve fixme-s in code

### PR DESCRIPTION
**Pull request by lauris71**
WE2-479

Signed-off-by: Lauris Kaplinski <lauris@raulwalter.com>

### CODE REVIEW

File: .github/workflows/cmake-windows.yml [hunk 1]
```diff
@@ -72,4 +72,4 @@ jobs:
           path: |
             build/src/app/*.msi
             build/src/app/*.exe
-            build/**/*.pdb
+            build/**/*.pdb
```
Comment (metsma) [inline .github/workflows/cmake-windows.yml:75]: Revert this change

File: src/controller/command-handlers/authenticate.cpp [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020-2024 Estonian Information System Authority
+ * Copyright (c) 2020-2023 Estonian Information System Authority
```
Comment (metsma) [inline src/controller/command-handlers/authenticate.cpp:2]:
```suggestion
 * Copyright (c) 2020-2024 Estonian Information System Authority
```

File: src/controller/command-handlers/authenticate.cpp [hunk 3]
```diff
@@ -51,20 +52,18 @@ QVariantMap createAuthenticationToken(const QString& signatureAlgorithm,
         {"format", QStringLiteral("web-eid:1.0")},
         {"appVersion",
          QStringLiteral("https://web-eid.eu/web-eid-app/releases/%1")
-             .arg(QApplication::applicationVersion())},
+             .arg(qApp->applicationVersion())},
```
Comment (metsma) [inline src/controller/command-handlers/authenticate.cpp:55]:
```suggestion
             .arg(QApplication::applicationVersion())},
```

File: src/controller/command-handlers/authenticate.cpp [hunk 4]
```diff
@@ -51,20 +52,18 @@ QVariantMap createAuthenticationToken(const QString& signatureAlgorithm,
...
 QByteArray createSignature(const QString& origin, const QString& challengeNonce,
                            const ElectronicID& eid, const pcsc_cpp::byte_vector& pin)
 {
-    static const std::map<JsonWebSignatureAlgorithm, QCryptographicHash::Algorithm>
-        SIGNATURE_ALGO_TO_HASH {
+    static const auto SIGNATURE_ALGO_TO_HASH =
+        std::map<JsonWebSignatureAlgorithm, QCryptographicHash::Algorithm> {
             {JsonWebSignatureAlgorithm::RS256, QCryptographicHash::Sha256},
             {JsonWebSignatureAlgorithm::PS256, QCryptographicHash::Sha256},
-            {JsonWebSignatureAlgorithm::ES256, QCryptographicHash::Sha256},
```
Comment (metsma) [inline src/controller/command-handlers/authenticate.cpp:65]: Revert these changes

File: src/controller/command-handlers/authenticate.cpp [hunk 5]
```diff
@@ -82,8 +81,8 @@ QByteArray createSignature(const QString& origin, const QString& challengeNonce,
     // The value that is signed is hash(origin)+hash(challenge).
     const auto hashToBeSignedQBytearray =
         QCryptographicHash::hash(originHash + challengeNonceHash, hashAlgo);
-    const pcsc_cpp::byte_vector hashToBeSigned {hashToBeSignedQBytearray.cbegin(),
-                                                hashToBeSignedQBytearray.cend()};
+    const auto hashToBeSigned =
+        pcsc_cpp::byte_vector {hashToBeSignedQBytearray.cbegin(), hashToBeSignedQBytearray.cend()};
```
Comment (metsma) [inline src/controller/command-handlers/authenticate.cpp:85]:
```suggestion
    const pcsc_cpp::byte_vector hashToBeSigned {hashToBeSignedQBytearray.cbegin(),
                                                hashToBeSignedQBytearray.cend()};
```

File: src/controller/command-handlers/authenticate.cpp [hunk 6]
```diff
@@ -97,12 +96,11 @@ QByteArray createSignature(const QString& origin, const QString& challengeNonce,
...
-    requireArgumentsAndOptionalLang(
-        {"challengeNonce", "origin"}, arguments,
-        R"("challengeNonce": "<challenge nonce>", "origin": "<origin URL>")");
+    requireArgumentsAndOptionalLang({"challengeNonce", "origin"}, arguments,
+                                    "\"challengeNonce\": \"<challenge nonce>\", "
+                                    "\"origin\": \"<origin URL>\"");
 
-    challengeNonce = validateAndGetArgument<decltype(challengeNonce)>(
-        QStringLiteral("challengeNonce"), arguments);
+    challengeNonce = validateAndGetArgument<QString>(QStringLiteral("challengeNonce"), arguments);
```
Comment (metsma) [inline src/controller/command-handlers/authenticate.cpp:103]: Revert

File: src/controller/command-handlers/authenticate.cpp [hunk 7]
```diff
@@ -122,26 +120,26 @@ QVariantMap Authenticate::onConfirm(WebEidUI* window,
...
 
     } catch (const VerifyPinFailed& failure) {
         switch (failure.status()) {
-        case VerifyPinFailed::Status::PIN_ENTRY_CANCEL:
-        case VerifyPinFailed::Status::PIN_ENTRY_TIMEOUT:
+        case electronic_id::VerifyPinFailed::Status::PIN_ENTRY_CANCEL:
+        case electronic_id::VerifyPinFailed::Status::PIN_ENTRY_TIMEOUT:
             break;
-        case VerifyPinFailed::Status::PIN_ENTRY_DISABLED:
+        case electronic_id::VerifyPinFailed::Status::PIN_ENTRY_DISABLED:
```
Comment (metsma) [inline src/controller/command-handlers/authenticate.cpp:142]: Revert

File: src/controller/command-handlers/sign.cpp [hunk 8]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020-2024 Estonian Information System Authority
+ * Copyright (c) 2020-2023 Estonian Information System Authority
```
Comment (metsma) [inline src/controller/command-handlers/sign.cpp:2]:
```suggestion
 * Copyright (c) 2020-2024 Estonian Information System Authority
```

File: src/controller/command-handlers/sign.cpp [hunk 9]
```diff
@@ -95,20 +96,21 @@ void Sign::emitCertificatesReady(const std::vector<CardCertificateAndPinInfo>& c
...
-        // Erase PIN memory.
-        // TODO: Use a scope guard. Verify that the buffers are actually zeroed
-        // and no copies remain.
-        std::fill(pin.begin(), pin.end(), '\0');
-
         return {{QStringLiteral("signature"), signature.first},
                 {QStringLiteral("signatureAlgorithm"), signature.second}};
 
     } catch (const VerifyPinFailed& failure) {
+
```
Comment (metsma) [inline src/controller/command-handlers/sign.cpp:113]:
```suggestion
```

File: src/controller/command-handlers/signauthutils.cpp [hunk 10]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020-2024 Estonian Information System Authority
+ * Copyright (c) 2020-2023 Estonian Information System Authority
```
Comment (metsma) [inline src/controller/command-handlers/signauthutils.cpp:2]:
```suggestion
 * Copyright (c) 2020-2024 Estonian Information System Authority
```

File: src/controller/command-handlers/signauthutils.hpp [hunk 11]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020-2024 Estonian Information System Authority
+ * Copyright (c) 2020-2023 Estonian Information System Authority
```
Comment (metsma) [inline src/controller/command-handlers/signauthutils.hpp:2]:
```suggestion
 * Copyright (c) 2020-2024 Estonian Information System Authority
```

File: src/controller/command-handlers/signauthutils.hpp [hunk 12]
```diff
@@ -22,17 +22,13 @@
...
 #include <QVariantMap>
 
 class WebEidUI;
-
-namespace electronic_id
-{
-class ElectronicID;
-class SignatureAlgorithm;
-} // namespace electronic_id
+class QSslCertificate;
```
Comment (metsma) [inline src/controller/command-handlers/signauthutils.hpp:31]: Revert

**Review by metsma — CHANGES_REQUESTED** [review]

**Review by metsma — APPROVED** [review]

**Review by mrts — CHANGES_REQUESTED** [review]

File: src/ui/webeiddialog.cpp [hunk 13]
```diff
@@ -287,7 +290,11 @@ QString WebEidDialog::getPin()
 {
     // getPin() is called from background threads and must be thread-safe.
     // QString uses QAtomicPointer internally and is thread-safe.
-    return pin;
+    // There should be only single reference and this is transferred to the caller for safety
+    QString ret = pin;
+    eraseData(pin);
+    pin.clear();
+    return ret;
```
Comment (metsma) [inline src/ui/webeiddialog.cpp:297]:
```suggestion
    // There should be only single reference and this is transferred to the caller for safety
    return std::move(pin);
```

Comment (metsma) [reply src/ui/webeiddialog.cpp:297]: I think this is better alternative

File: src/ui/webeiddialog.cpp [hunk 14]
```diff
@@ -20,6 +20,8 @@
  * SOFTWARE.
  */
 
+#include "utils/erasedata.hpp"
+
```
Comment (metsma) [inline src/ui/webeiddialog.cpp:24]:
```suggestion
```

Comment (metsma) [reply src/ui/webeiddialog.cpp:24]: See next comment

File: src/controller/command-handlers/signauthutils.hpp [hunk 15]
```diff
@@ -45,6 +45,6 @@ extern template QString validateAndGetArgument<QString>(const QString& argName,
 extern template QByteArray
 validateAndGetArgument<QByteArray>(const QString& argName, const QVariantMap& args, bool allowNull);
 
-pcsc_cpp::byte_vector getPin(const electronic_id::ElectronicID& card, WebEidUI* window);
+void getPin(pcsc_cpp::byte_vector& pin, const electronic_id::ElectronicID& eid, WebEidUI* window);
```
Comment (metsma) [inline src/controller/command-handlers/signauthutils.hpp:48]: Why this is needed? Result is moved out from function

**Review by metsma — CHANGES_REQUESTED** [review]

File: src/controller/command-handlers/signauthutils.hpp (hunk 15 above)
Comment (lauris71) [reply src/controller/command-handlers/signauthutils.hpp:48]: Cleaner semantics: The physical memory where PIN resides is always owned by caller (and thus the copying problem simply does not exist)
AFAIK c++ std::move does not guarantee no-copy rule:
https://stackoverflow.com/questions/70078468/are-move-semantics-guaranteed-by-the-standard

File: src/ui/webeiddialog.cpp (hunk 13 above)
Comment (lauris71) [reply src/ui/webeiddialog.cpp:297]: AFAIK std::move gives only hint, not guarantee, of non-duplication of the data.
All sane implementations do the right thing, but IMO we should rely on specification, not implementation.
I may be wrong, but then I'd say that the specification is confusing, and that is not a good thing for security critical code

**Review by metsma — APPROVED** [review]
